arch/sim/src/sim/up_interruptcontext.c

```diff
@@ -51,6 +51,7 @@ bool up_interrupt_context(void)
  * Name: up_doirq
  ****************************************************************************/
 
+__attribute__ ((visibility("default")))
```

Comment:
why need

Reply:
Otherwise native linker was not able to find up_doirq() in final link step. There's another place in up_head() for main(). It's overwriting fvisibility=hidden flag. Not sure, if there's a better solution possible.

Reply:
Let me check, if there's a better solution possible w/o adding this to the code.

Reply:
Ok.

Reply:
I did some further tests on Linux ARM64 and MacOS AARCH64. Without the attribute up_doirq() is not found on MacOS AARCH64 while it's still found on Linux ARM64. In both cases it's compiled with 'fvisibility=hidden" and objdump on Linux ARM shows, that up_doirq in marked ".hidden". On MacOS objdump doesn't provide this information. My current conclusion is, that it should not link on Linux ARM64, too. But in reality it does.

Reply:
Using an x86_64 based Macbook, sim:nsh is not working, too. It fails linking at the same two functions: up_doirq() / main(). This nails the issue down to be most likely an clang related issue regardless of host CPU architecture. Are there any simulations known to work using clang based compiler suite?

Reply:
do_irq is used by nuttx elf self, not to other so, it's strange that visibility could impact the linking of do_irq.

Reply:
In fact up_doirq() is also called from up_handle_irq() in up_hostirq.c, too. Removing the attribute, results in the following link error:

    LD:  nuttx
    Undefined symbols for architecture arm64:
      "_up_doirq", referenced from:
          _up_handle_irq in up_hostirq.o
    ld: symbol(s) not found for architecture arm64

The linker is invoked:

    "cc" -g -fomit-frame-pointer -L board  -o /Users/piet/Projects/NuttX/feature-sim-m1-support/nuttx/nuttx  nuttx.rel up_hostirq.o up_hostmemory.o up_hosttime.o up_simuart.o up_hostmisc.o -lpthread -lz

The missing symbol up_doirq() is part of nuttx.rel object file, which is build with the following command:

    ld -r  -L"/Users/piet/Projects/NuttX/feature-sim-m1-support/nuttx/staging" -L board  -o nuttx.rel up_head.o  -lsched -ldrivers -lboards -lc -lmm -larch -lapps -lfs -lbinfmt -lboard  

This partially linking steps hides up_doirq() for the final link step. And this seems be different on none MacOS based host system.

I would be glad, if there's a better solution w/o the attribute in the source code itself.

Reply:
Thanks for the detailed explanation, the need of visibility attribute is clear to me now.